### PR TITLE
handling shut valve

### PR DESCRIPTION
**Pull request by GitPaean**
When a valve is SHUT, the rate and pressure drop for that segment will be zero.

### CODE REVIEW

File: opm/simulators/wells/MultisegmentWellEval.cpp [hunk 1]
```diff
@@ -1554,6 +1554,18 @@ assembleICDPressureEq(const int seg,
     // top segment can not be a spiral ICD device
     assert(seg != 0);
 
+    if(this->segmentSet()[seg].segmentType() == Segment::SegmentType::VALVE) {
+        const Valve& valve = this->segmentSet()[seg].valve();
+        if (valve.status() == Opm::ICDStatus::SHUT) { // we switch here to be a zero rate equation
+            resWell_[seg][SPres] = this->primary_variables_evaluation_[seg][GTotal].value();
+            duneD_[seg][seg][SPres][GTotal] = 1.;
+
+            auto& ws = well_state.well(baseif_.indexOfWell());
```
Comment (GitPaean) [inline opm/simulators/wells/MultisegmentWellEval.cpp:1563]: If the valve is shut and there no active perforations for this shut branch, there might be single matrix here, similar to a closed well without active perforations.  There might need to be more development to handle that special situation.

File: opm/simulators/wells/MultisegmentWellEval.cpp [hunk 2]
```diff
@@ -1554,6 +1554,18 @@ assembleICDPressureEq(const int seg,
     // top segment can not be a spiral ICD device
     assert(seg != 0);
 
+    if(this->segmentSet()[seg].segmentType() == Segment::SegmentType::VALVE) {
+        const Valve& valve = this->segmentSet()[seg].valve();
+        if (valve.status() == Opm::ICDStatus::SHUT) { // we switch here to be a zero rate equation
```
Comment (bska) [inline opm/simulators/wells/MultisegmentWellEval.cpp:1559]: For what it's worth you can reduce the nesting here if you use C++17's `if` with an initialisation statement
```C++
if (const auto& segment = this->segmentSet()[seg];
    (segment.segmentType() == Segment::SegmentType::VALVE) &&
    (segment.valve().status() == Opm::ICDStatus::SHUT))
{
    // Handle shut valve
}
```

Comment (GitPaean) [reply opm/simulators/wells/MultisegmentWellEval.cpp:1559]: Will do that when the result is confirmed.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by totto82** [comment]
Thanks I will test it asap.

**Comment by totto82** [comment]
Testing done. From my point of view this can be merged. Thanks. I will close the other PR

**Comment by GitPaean** [comment]
jenkins build this update_data please

File: tests/update_reference_data.sh [hunk 3]
```diff
@@ -154,6 +154,7 @@ tests[udq_pyaction]="flow udq_actionx PYACTION_WCONPROD"
 tests[spe1_foam]="flow spe1_foam SPE1FOAM"
 tests[wsegsicd]="flow wsegsicd TEST_WSEGSICD"
 tests[wsegaicd]="flow wsegaicd BASE_MSW_WSEGAICD"
+tests[wsegaicd]="flow wsegvalv BASE_MSW_WSEGVALV"
```
Comment (bska) [inline tests/update_reference_data.sh:157]: This should **probably** say `"wsegvalv"` in order to match the test name.

Comment (GitPaean) [reply tests/update_reference_data.sh]: Thanks.

**Comment by GitPaean** [comment]
jenkins build this update_data please

File: opm/simulators/wells/MultisegmentWellEval.cpp (hunk 2 above)
Comment (GitPaean) [reply opm/simulators/wells/MultisegmentWellEval.cpp:1559]: I see the reason now. I should have deleted the extra `}`.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=778 please

**Comment by GitPaean** [comment]
I think it **can** be the proper way to handle `SHUT` valve, which is also giving desired results following the test report from @totto82 . It still has the limitation not being able to handle the `DEAD` shut branch without active connections, which will be more involved to tangle, but we can tangle it when we have the case for it. The idea is relatively clear, while might be more development.

The jenkins gives green now. Maybe @totto82  or @bska can press green if there is no more comments.

**Comment by totto82** [comment]
Thanks @GitPaean for your help on this!

**Pull request merged**